Add utility functions to compare schemas

sam-swarr opened the pull request:
This PR introduces some checks that can be used to help verify if changes to a schema are "safe". The idea is that you generate a `GraphQLSchema` object that represents the schema before and after a schema change (for example by using `buildClientSchema`). You can then pass the `oldSchema` and `newSchema` into these utility functions. The functions each check for different types of breaking changes, i.e. changes in the `newSchema` that are not backwards compatible with `oldSchema`. They can be used to help prevent regressions caused by unsafe schema changes that would break older versions of clients.

leebyron commented:
This is awesome!

My high level feedback is that we should probably have a single function like `findBreakingChanges` which is capable of returning a full list of described breakages.

Otherwise it's up to you to call all of these functions in succession which could require some boilerplate.

sam-swarr commented:
> My high level feedback is that we should probably have a single function like findBreakingChanges which is capable of returning a full list of described breakages.

Good call. I added a top level `findBreakingChanges` check and added a test case.

leebyron commented on src/utilities/schemaComparisons.js:
```diff
@@ -0,0 +1,220 @@
+    if (!{}.hasOwnProperty.call(oldTypeMap,typeName)) {
+      continue;
+    }
+    const oldType:GraphQLType = oldTypeMap[typeName];
```
nit: space after `:`

But also, you don't need to add types to variables you create - Flow infers all of that for you. Only need to annotate function args & returns. You can remove all of them and Flow will keep on chugging.

sam-swarr replied:
Ah good to know. I removed type declarations for all local variables.

leebyron commented on src/utilities/schemaComparisons.js:
```diff
@@ -0,0 +1,220 @@
+  oldSchema: GraphQLSchema,
+  newSchema: GraphQLSchema
+): Set<string> {
+  return new Set([
```
What's the reasoning behind using Sets throughout instead of just Arrays?

It seems like there ends up being a lot of work converting between Sets and Arrays and Object maps and we should ensure it's worth doing all that.

Also, I bet users of this function will expect an array of errors rather than a Set.

It's cool to use Set as an implementation detail in deriving these breakages though!

e.g. this could become:

```js
return [
  ...findRemovedTypes(oldSchema, newSchema),
  ...findTypesThatChangedType(oldSchema, newSchema),
  ...findBreakingFieldChanges(oldSchema, newSchema),
  ...findTypesRemovedFromUnions(oldSchema, newSchema),
  ...findValuesRemovedFromEnums(oldSchema, newSchema)
];
```

leebyron commented on src/utilities/index.js:
```diff
@@ -60,3 +60,13 @@ export {
+
+// Helpful functions to compare two GraphQLSchemas and detect breaking changes.
+export {
+  findBreakingChanges,
```
You probably only need to export this function (same below). All the other functions are really just implementation details of this one.

sam-swarr replied:
I kept the other functions exported for a couple of reasons. One reason is so that the unit test can test them individually. The other is that other users might be interested in only a subset of the checks. This way they can pick and choose.

leebyron replied:
Unit tests should be importing from that file directly - not from this package module.

I don't think there's a good reason to only partially check for breaking changes? Perhaps if you want to filter out certain breakages a better way would be to filter based on data in the BreakingChange objects instead of manually assembling these functions yourself.

If there real need in the future then we could export these then - otherwise I'd prefer to keep the surface area of the exported API smaller and preserve the option value to change the implementation of `findRemovedTypes` in the future.

sam-swarr replied:
> Unit tests should be importing from that file directly - not from this package module.

Turns out the test was already directly importing from the file.

> If there real need in the future then we could export these then - otherwise I'd prefer to keep the surface area of the exported API smaller and preserve the option value to change the implementation of findRemovedTypes in the future.

That's a very good point. Updated so that only `findBreakingChanges` is exported.

leebyron commented on src/utilities/schemaComparisons.js:
```diff
@@ -0,0 +1,220 @@
+* Given two schemas, returns a Set containing descriptions of any breaking
+* changes in the newSchema related to changing the type of a type.
+*/
+export function findTypesThatChangedType(
```
A better name for this might be `findTypesThatChangedKind`

sam-swarr replied:
Yep, I like that better.

leebyron commented on src/utilities/schemaComparisons.js:
```diff
@@ -0,0 +1,220 @@
+* changes in the newSchema related to the fields on a type. This includes if
+* a field has been removed from a type or if a field has changed type.
+*/
+export function findBreakingFieldChanges(
```
`findFieldsThatChangedType` might be a better name relative to the names of the other functions

sam-swarr replied:
Updated

leebyron commented on src/utilities/schemaComparisons.js:
```diff
@@ -0,0 +1,220 @@
+  const newTypeMap: TypeMap = newSchema.getTypeMap();
+
+  const typesRemovedFromUnion: Array<string> = [];
+  for (const typeName: string in oldTypeMap) {
```
Use `Object.keys(oldTypeMap).forEach(typeName =>` instead.

sam-swarr replied:
Converted all for in loops to forEach.

leebyron commented on src/utilities/schemaComparisons.js:
```diff
@@ -0,0 +1,220 @@
+
+  const typesRemovedFromUnion: Array<string> = [];
+  for (const typeName: string in oldTypeMap) {
+    if (!{}.hasOwnProperty.call(oldTypeMap,typeName)) {
```
this will be removed by the suggestion above, but in the future, this sort of check is better written:

```js
oldTypeMap.hasOwnProperty(typeName)
```

sam-swarr replied:
I had that, but then I got `TypeError: oldTypeMap.hasOwnProperty is not a function`. I guess Flow-typed objects don't inherit `hasOwnProperty`?

Doesn't really matter though since I took your advice above and converted to forEach loops.

leebyron commented on src/utilities/schemaComparisons.js:
```diff
@@ -0,0 +1,220 @@
+  const newTypeMap: TypeMap = newSchema.getTypeMap();
+
+  const valuesRemovedFromEnums: Array<string> = [];
+  for (const typeName: string in oldTypeMap) {
```
same

leebyron commented:
Should this, rather than returning a list of strings, return a list of some kind of object, where `description` is one field, but another field is some constant that tells a machine what kind of breakage it is?

sam-swarr commented:
> Should this, rather than returning a list of strings, return a list of some kind of object, where description is one field, but another field is some constant that tells a machine what kind of breakage it is?
Good idea. I added a `BreakingChange` object that is now returned by these functions. That should allow the caller more flexibility in handling the results.

leebyron commented on src/utilities/schemaComparisons.js:
```diff
@@ -0,0 +1,239 @@
+  VALUE_REMOVED_FROM_ENUM: 'VALUE_REMOVED_FROM_ENUM',
+};
+
+export type BreakingChange = {
```
nice!

leebyron commented on src/utilities/schemaComparisons.js:
```diff
@@ -0,0 +1,239 @@
+  const oldTypes = Object.keys(oldSchema.getTypeMap());
+  const newTypes = new Set(Object.keys(newSchema.getTypeMap()));
+  return oldTypes.filter(typeName => !newTypes.has(typeName)).map(
+    type => {
```
this arrow only has one statement, a return statement, which means it can be re-written as:

```
map(type => ({
  type: BreakingChangeType.TYPE_REMOVED,
  description: `${type} was removed`,
}))
```

note the wrapping `({})` around the object literal to disambiguate from `{}` which arrows treat as a statement body :/

leebyron commented on src/utilities/schemaComparisons.js:
```diff
@@ -0,0 +1,239 @@
+    const oldType = oldTypeMap[typeName];
+    const newType = newTypeMap[typeName];
+    if (!(oldType instanceof newType.constructor)) {
+      typesThatChangedType.push(
```
`typesThatChangedKind`

or perhaps just `breakingChanges`?

leebyron commented on src/utilities/schemaComparisons.js:
```diff
@@ -0,0 +1,239 @@
+        {
+          type: BreakingChangeType.TYPE_CHANGED_KIND,
+          description: `${typeName} changed from a ` +
+            `${oldType.constructor.name} to a ${newType.constructor.name}`,
```
These error names might be less clear to non JS users if this utility is used along with other servers.

This might print:

`Foo changed from a GraphQLObjectType to a GraphQLScalarType`

but might be easier to understand as:

`Foo changed from an Object type to a Scalar type`

sam-swarr replied:
Updated so that the more generic Object/Scalar/Interface term is used.

igorcanadi commented:
Awesome to see this come together!

leebyron commented on src/utilities/schemaComparisons.js:
```diff
@@ -0,0 +1,244 @@
+): Array<BreakingChange> {
+  const oldTypes = Object.keys(oldSchema.getTypeMap());
+  const newTypes = new Set(Object.keys(newSchema.getTypeMap()));
+  return oldTypes.filter(typeName => !newTypes.has(typeName)).map(
```
No need to convert object to set:

```js
const oldTypeNames = Object.keys(oldSchema.getTypeMap());
const newTypes = newSchema.getTypeMap();
return oldTypeNames.filter(typeName => !newTypes[typeName]).map(
```

sam-swarr replied:
Removed Set.

leebyron commented on src/utilities/schemaComparisons.js:
```diff
@@ -0,0 +1,244 @@
+  return breakingChanges;
+}
+
+function prettifyTypeKind(typeKind: string): string {
```
lol this is kind of amazing - was expecting something more like `if (x instanceof GraphQLEnumType) { return "an Enum"; }` but I'm cool with this.

denvned replied:
I think just adding a `prettyName` static property to each graphql type class would comply the open-closed principle even better. It could be accessed as `oldType.constructor.prettyName`.

sam-swarr replied:
Updated to use the new static `prettyName` property.

leebyron commented:
Nice work

sam-swarr commented:
Thanks for the thorough review!
